### PR TITLE
Add systemd analyze test

### PR DESCRIPTION
**Pull request by 7c6f434c**
###### Motivation for this change

Trying to improve the NixOS test infrastructure; fixing the handling of shared dir in Python test driver, adding a `copy_from_vm` function, and adding a test exporting evereything `systemd-analyze` has to say.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli

### CODE REVIEW

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
...
                 raise Exception("Cannot convert screenshot")
 
+    def copy_from_vm(self, source: str, target_dir: str = "") -> None:
+        """Copy a file from the VM (specified by an in-VM source path) to $out
+        (with an optional subdirectory parameter). The file is copied via the
+        shared directory.
+        """
+        # Compute the source, target, and intermediate shared file names
+        out_dir = os.environ.get("out", os.getcwd())
+        filename = re.sub(r".*/", "", source)
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: Please use `pathlib` here to calculate basenames and assemble paths:

https://docs.python.org/3/library/pathlib.html#methods-and-properties

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
...
+        int_shared_dir = "/tmp/xchg"
+        shared_temp = tempfile.mkdtemp(dir=self.shared_dir)
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: This should use [`tempfile.TemporaryDirectory`](https://devdocs.io/python~3.7/library/tempfile#tempfile.TemporaryDirectory) as a context manager, it also takes care of cleaning up afterwards.

File: nixos/lib/test-driver/test-driver.py [hunk 3]
```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
...
+        shared_temp_basename = re.sub(r".*/", "", shared_temp)
+        shared_temp_in = "{}/{}".format(int_shared_dir, shared_temp_basename)
+        src_out = "{}/{}".format(shared_temp, filename)
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: You can join `pathlib` paths by just `/` in between them.

File: nixos/lib/test-driver/test-driver.py [hunk 4]
```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
...
+        src_in = "{}/{}".format(shared_temp_in, filename)
+        tgt = "{}/{}/".format(out_dir, target_dir)
+        # Copy the file to the shared directory inside VM
+        if not self.is_up():
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: current behaviour in the `Machine` class is to start machines if you request things from them, so we should do here, too.

File: nixos/lib/test-driver/test-driver.py [hunk 5]
```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
...
+            raise Exception("The machine needs to be up for copying")
+        self.succeed("mkdir -p {}".format(shlex.quote(shared_temp_in)))
+        self.succeed("cp -r {} {}".format(shlex.quote(source), shlex.quote(src_in)))
+        self.succeed("sync")
+        # Copy the file from the shared directory outside VM
+        ret = subprocess.run(["mkdir", "-p", tgt])
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: This probably can also be done with python stdlib instead of shelling out to `mkdir`.

File: nixos/tests/all-tests.nix [hunk 6]
```diff
@@ -260,6 +260,7 @@ in
   syncthing-init = handleTest ./syncthing-init.nix {};
   syncthing-relay = handleTest ./syncthing-relay.nix {};
   systemd = handleTest ./systemd.nix {};
+  systemd-analyze = handleTest ./systemd-analyze.nix {};
```
Comment (flokli) [inline nixos/tests/all-tests.nix]: It'd be nice if we could split this into a separate PR, but fine to keep it around for now until the test driver additions are done.

**Comment by 7c6f434c** [comment]
Depends on/includes #74077

File: nixos/tests/systemd-analyze.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    { pkgs, lib, ... }:
+    { boot.kernelPackages = lib.mkIf latestKernel pkgs.linuxPackages_latest;
+      sound.enable = true; # needed for the factl test, /dev/snd/* exists without them but udev doesn't care then
+    };
+
+  testScript = ''
+    machine.wait_for_unit("multi-user.target")
+
+    with subtest("Prepare output dir"):
+        machine.succeed("mkdir systemd-analyze")
```
Comment (flokli) [inline nixos/tests/systemd-analyze.nix]: I think we can put the equivalent of doing a `mkdir -p $(dirname $dst)` into the`copy_from_vm` function, so we don't have to repeat ourselves.

Also see `nixos/tests/blivet.nix`, which would benefit from that too (if it wouldn't have been broken, see https://github.com/NixOS/nixpkgs/issues/33496).

Comment (flokli) [reply nixos/tests/systemd-analyze.nix]: This would be a change in #74077 of course, I can just better explain it with the example here.

File: nixos/tests/systemd-analyze.nix [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+
+    def run_systemd_analyze(args, name):
+        tgt_dir = "systemd-analyze"
+        machine.succeed(
+            "systemd-analyze {} > {}/{} 2> {}/{}.err".format(
+                " ".join(args), tgt_dir, name, tgt_dir, name
+            )
```
Comment (flokli) [inline nixos/tests/systemd-analyze.nix]: This seems to be overly generic. We only call it with one argument.

What about just executing `machine.succeed(…)` with a list of the "raw" commands:

```
machine.succeed(
    "systemd-analyze blame > blame.txt",
    "systemd-analyze critical-chain > critical-chain.txt",
    "systemd-analyze dot > dependencies.dot",
    "systemd-analyze plot > systemd-analyze.svg",
)

File: nixos/tests/systemd-analyze.nix [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        )
+
+
+    with subtest("Print statistics"):
+        run_systemd_analyze(["blame"], "blame.txt")
+        run_systemd_analyze(["critical-chain"], "critical-chain.txt")
+        run_systemd_analyze(["dot"], "dependencies.dot")
+        run_systemd_analyze(["plot"], "systemd-analyze.svg")
+
+    with subtest("Install the resulting data"):
```
Comment (flokli) [inline nixos/tests/systemd-analyze.nix]: with subtest("Copy the resulting data into $out"):

File: nixos/tests/systemd-analyze.nix [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        machine.copy_from_vm("systemd-analyze", "")
+        machine.copy_from_vm("systemd-analyze/systemd-analyze.svg", "")
```
Comment (flokli) [inline nixos/tests/systemd-analyze.nix]: Some files seem to be missing here.

**Comment by 7c6f434c** [comment]
>> +            "systemd-analyze {} > {}/{} 2> {}/{}.err".format(
>+                " ".join(args), tgt_dir, name, tgt_dir, name
>+            )
>
>This seems to be overly generic. We only call it with one argument.

True, but cheap

>What about just executing `machine.succeed(…)` with a list of the "raw" commands:
>
>```
>machine.succeed(
>    "systemd-analyze blame > blame.txt",
>    "systemd-analyze critical-chain > critical-chain.txt",
>    "systemd-analyze dot > dependencies.dot",
>    "systemd-analyze plot > systemd-analyze.svg",
>)

Well, not exactly; stderr from `dot` subcommand is actually useful in some sense. And also I do want to put stuff into a subdirectory that I can later copy as a whole.

**Comment by 7c6f434c** [comment]
>+    with subtest("Prepare output dir"):
>+        machine.succeed("mkdir systemd-analyze")
>
>I think we can put the equivalent of doing a `mkdir -p $(dirname $dst)` into the`copy_from_vm` function, so we don't have to repeat ourselves.

This is in-VM directory for `systemd-analyze` output that will later become the original source location for `copy_from_vm`

>Also see `nixos/tests/blivet.nix`, which would benefit from that too (if it wouldn't have been broken, see https://github.com/NixOS/nixpkgs/issues/33496).

**Comment by 7c6f434c** [comment]
>+    with subtest("Install the resulting data"):
>+        machine.copy_from_vm("systemd-analyze", "")
>+        machine.copy_from_vm("systemd-analyze/systemd-analyze.svg", "")
>
>Some files seem to be missing here.

No: I copy The Main Picture to the toplevel, and then the entire set of data as a subdirectory.

**Comment by flokli** [comment]
With #74077 merged, can this be rebased on top of master?

**Comment by 7c6f434c** [comment]
OK, the hard part of rebase (building the kernel from `master` which is ahead of Hydra) is done, the test still seems to work (of course it was always rebased on #74077 so no surprise here)

**Comment by 7c6f434c** [comment]
Maybe I should re-mention @flokli

**Comment by flokli** [comment]
Can you move the replies sent via E-Mail into the thread? GitHub doesn't properly attach these, and keeping the discussion in threads seems like a good thing to do.

**Comment by 7c6f434c** [comment]
Both options provided by GitHub (threaded and single-stream) are annoying in different ways, of course (which is why mailing lists, which are actually suitable for discussion, allow switching the navigation mode), but oh well… got around to starting a Firefox instance.

File: nixos/tests/systemd-analyze.nix (hunk 8 above)
Comment (7c6f434c) [reply nixos/tests/systemd-analyze.nix]: > overly generic 

True, but cheap

> list of the "raw" commands

The replacement list loses stderr of the `dot` subcommand which does contain some data that could be in principle useful.

Also, I do want to have a subdirectory that I can copy as a whole.

File: nixos/tests/systemd-analyze.nix (hunk 10 above)
Comment (7c6f434c) [reply nixos/tests/systemd-analyze.nix]: No: I copy The Main Picture to the toplevel, and then the entire set of data as a subdirectory.

File: nixos/tests/systemd-analyze.nix (hunk 7 above)
Comment (7c6f434c) [reply nixos/tests/systemd-analyze.nix]: This is in-VM directory for `systemd-analyze` output that will later become the original source location for `copy_from_vm`

File: nixos/tests/systemd-analyze.nix (hunk 8 above)
Comment (flokli) [reply nixos/tests/systemd-analyze.nix]: I don't care that much. I think it might be more readable if written more explicitly, but your call.

File: nixos/tests/systemd-analyze.nix (hunk 10 above)
Comment (flokli) [reply nixos/tests/systemd-analyze.nix]: In that case, more comments and less abstractions would help understanding it ;-)

File: nixos/tests/systemd-analyze.nix (hunk 8 above)
Comment (7c6f434c) [reply nixos/tests/systemd-analyze.nix]: Tried. It becomes pretty Black-fragile. Agree on the point of more comments needed.

File: nixos/tests/systemd-analyze.nix (hunk 10 above)
Comment (7c6f434c) [reply nixos/tests/systemd-analyze.nix]: Less asbtractions mean more Black pain. More comments added

**Comment by flokli** [comment]
@GrahamcOfBorg build systemd-analyze

**Comment by flokli** [comment]
@GrahamcOfBorg test systemd-analyze

**Comment by flokli** [comment]
Thanks!

**Comment by 7c6f434c** [comment]
Thanks for the carefuk rereading of all that and for the explanations!

**Comment by flokli** [comment]
Thanks for staying through till the end, and enduring my pickyness :laughing:

**Comment by 7c6f434c** [comment]
I have to recognise that all of the pickyness was clearly actionable.
